### PR TITLE
mixedversion: introduce deployment modes

### PR DESCRIPTION
**Pull request by renatolabs**
This commit adds the notion of a "deployment mode" to mixedversion tests. Currently, we only define a `system-only` deployment mode, corresponding to the deployment performed on all mixedversion tests to date: cockroach is started without any tenants (virtual clusters), and client applications talk directly to the "system interface". This commit adds the functionality necessary for tests to be able to select what deployment modes make sense for them, if they verify something specific to some deployment mode.

In the future, more deployment modes will be added: tenants in shared-process mode, and external process (serverless-style) tenants.

Epic: none

Release note: None

### CODE REVIEW

**Review by herkolategan — APPROVED** [review]
Nice!

**Review by DarrylWong — APPROVED** [review]
Looks great!

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go [hunk 1]
```diff
@@ -441,7 +444,7 @@ func Test_stepSelectorFilter(t *testing.T) {
 			name:                   "no filter",
 			predicate:              func(*singleStep) bool { return true },
 			expectedAllSteps:       true,
-			expectedRandomStepType: runHookStep{},
+			expectedRandomStepType: restartWithNewBinaryStep{},
```
[the remark this answers is not in the document]
Comment (srosenberg) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go:447]: It's rather subtle but prng determinism corresponds to both `seed` (which hasn't changed) _and_ the number of calls to prng. The latter did indeed change; on L630, we've added a new call to prng.

**Review by srosenberg — APPROVED** [review]

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (renatolabs) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go:447]: Correct!

**Comment by renatolabs** [comment]
TFTRs!

bors r=herkolategan,DarrylWong,srosenberg

**Comment by renatolabs** [comment]
blathers backport 24.1 23.2 23.1

**Pull request merged**